File: webapp/backend/app/graph_connectivity.py
"""Validate that an attack-flow graph is one connected piece (when multiple nodes exist)."""

from __future__ import annotations

from typing import Any


def validate_attack_flow_connectivity(
    nodes: list[dict[str, Any]],
    edges: list[dict[str, Any]],
) -> tuple[bool, str]:
    """Return (True, "") if the flow may be shown; (False, user_message) if disconnected.

    Rules:
    - 0 or 1 node: always OK (nothing to connect).
    - 2+ nodes with no edges: not OK.
    - 2+ nodes with edges but more than one undirected connected component: not OK.
    """
    ids = {str(n["id"]) for n in nodes if n.get("id") not in (None, "")}
    if len(ids) <= 1:
        return True, ""

    if not edges:
        return False, (
            "The analysis found multiple entities but no relationships between them. "
            "An attack flow must show how steps link together. "
            "Try richer threat-report text, explicit procedures, or a STIX bundle that includes relationships, then analyze again."
        )

    adj: dict[str, set[str]] = {i: set() for i in ids}
    for e in edges:
        s, t = e.get("source"), e.get("target")
        if s is None or t is None:
            continue
        s, t = str(s), str(t)
        if s in ids and t in ids:
            adj[s].add(t)
            adj[t].add(s)

    visited: set[str] = set()
    components = 0
    for nid in ids:
        if nid in visited:
            continue
        components += 1
        stack = [nid]
        while stack:
            u = stack.pop()
            if u in visited:
                continue
            visited.add(u)
            for v in adj.get(u, ()):
                if v not in visited:
                    stack.append(v)

    if components > 1:
        return False, (
            "The graph is not fully connected: some parts of the attack have no path to the rest. "
            "This app expects a single linked attack flow. "
            "Use source material that ties techniques together, or one coherent incident narrative, then analyze again."
        )

    return True, ""

File: webapp/frontend/src/utils/graphConnectivity.ts
/** Same rules as backend `graph_connectivity.validate_attack_flow_connectivity`. */

export function isAttackFlowConnected(
  nodes: Array<{ id: string }>,
  edges: Array<{ source: string; target: string }>,
): boolean {
  const ids = new Set(
    nodes.map((n) => n.id).filter((id) => id != null && id !== ''),
  );
  if (ids.size <= 1) return true;
  if (!edges.length) return false;

  const adj = new Map<string, Set<string>>();
  ids.forEach((id) => adj.set(id, new Set()));
  for (const e of edges) {
    const s = String(e.source);
    const t = String(e.target);
    if (ids.has(s) && ids.has(t)) {
      adj.get(s)!.add(t);
      adj.get(t)!.add(s);
    }
  }

  const visited = new Set<string>();
  let components = 0;
  for (const id of ids) {
    if (visited.has(id)) continue;
    components += 1;
    const stack = [id];
    while (stack.length) {
      const u = stack.pop()!;
      if (visited.has(u)) continue;
      visited.add(u);
      for (const v of adj.get(u) ?? []) {
        if (!visited.has(v)) stack.push(v);
      }
    }
  }
  return components <= 1;
}
